### ISSUE TITLE
:set guifont does not correctly set font

### ISSUE THREAD

**Issue by jsouthwo** [vim/vim#1723]
I'm using [Source Code Pro](https://github.com/adobe-fonts/source-code-pro) from Adobe on Windows 7 in GVIM 7.4, and prefer the ExtraLight version of the font.

If I do `:set guifont=*` and select `Source Code Pro` as the font and `ExtraLight` as the font style, everything looks fine. The output of `:set guifont?` is `guifont=Source_Code_Pro_ExtraLight:h14:cANSI`.

However, if I put `guifont=Source_Code_Pro_ExtraLight:h14:cANSI` in my `vimrc` or run it within GVIM, the font displayed is Source Code Pro Regular, not ExtraLight. The output of `:set guifont?` is still `guifont=Source_Code_Pro_ExtraLight:h14:cANSI`.

I have tried it for the other styles in the font and they all work.

**Comment by mattn** [comment]
I don't reproduce this.

![image](https://user-images.githubusercontent.com/10111/46646306-8cbc9d80-cbc4-11e8-8eb3-47a3b9778c60.png)

**Comment by tonymec** [comment]
Does anything change if you put in your `vimrc` (or `gvimrc`)
```vim
set guifont=Source_Code_Pro_ExtraLight:h14:cDEFAULT
```
instead of `…:cANSI`? And/or if you try other font sizes? (Which ones to try is a question of trial and error.)

Another way to set the font in gvim is by using
```vim
hi Normal font=Source_Code_Pro_ExtraLight:h14:cDEFAULT
```
(see `:help highlight-font`). It *should* give the same result as setting `'guifont'` — but does it?

Best regards,
Tony.

**Comment by jsouthwo** [comment]
These all give the same result.

**Comment by ghost** [comment]
I encountered the same problem **on Win10**, when I wrote "set guifont=Consolas:h16:cDEFAULT" to _vimrc file, then I restart gvim, the font size still do not change.

**Comment by tonymec** [comment]
> I encountered the same problem **on Win10**, when I wrote "set guifont=Consolas:h16:cDEFAULT" to _vimrc file, then I restart gvim, the font size still do not change.

- Can you set the same font manually?
- If you can, then I don't know. If you cannot, then my guess is that the Consolas font either is not installed, or is installed but does not declare itself as fixed-width. (Only the GTK2 and GTK3 GUIs, neither of which is available on Windows, will accept to set a 'guifont' which does not declare itself as fixed-width in its metadata — and if the font is indeed variable-width, the result is usually ugly, because narrow glyphs such as lowercase i and l look too far apart and wide glyphs such as lowercase m and w seem too close together).

Best regards,
Tony.

**Comment by machintosher** [comment]
The problem here is that get_logfont in os_mswin.c populates the LOGFONTW structure with some defaults including the lfWeight. Subsequently, with it passes this LOGFONTW to get_font_handle and then CreateFontIndirectW Windows picks the font that matches the weight rather than the facename (lfFaceName). The problem is that there is no easy way to specify the weight. Specifying a 'b' for the font sets the weight to FW_BOLD but there is no way to precisely indicate the weight. Maybe the 'W' attribute can be used to specify the weight.

I've added a pull request that seeks to address this issue.

**Comment by k-takata** [comment]
This should be fixed by #4309. Closing.
